File: Algorithms/Easy/SquaresOfASortedArray.go
package easy

// Time complexity: O(n)
// Space complexity: O(1)
func sortedSquares_1(nums []int) []int {
	n := len(nums)
	res := make([]int, n)

	l, r := 0, n-1
	for i := n - 1; i > -1; i-- {
		if nums[l]*nums[l] > nums[r]*nums[r] {
			res[i] = nums[l] * nums[l]
			l++
		} else {
			res[i] = nums[r] * nums[r]
			r--
		}
	}

	return res
}

// Time complexity: O(n)
// Space complexity: O(1)
func sortedSquares_2(nums []int) []int {
	n := len(nums)
	res := make([]int, 0, n)

	// Obtaining the last element less than or equal to 0 using binary search
	l, r := 0, n-1
	for l < r {
		m := (l + r + 1) / 2
		if nums[m] <= 0 {
			l = m
		} else {
			r = m - 1
		}
	}
	r = l + 1

	for i := range nums {
		nums[i] *= nums[i]
	}

	for l > -1 && r < n {
		if nums[l] < nums[r] {
			res = append(res, nums[l])
			l--
		} else {
			res = append(res, nums[r])
			r++
		}
	}

	for l > -1 {
		res = append(res, nums[l])
		l--
	}

	for r < n {
		res = append(res, nums[r])
		r++
	}

	return res
}

File: Algorithms/Easy/977.Squares-Of-A-Sorted-Array/SquaresOfASortedArray.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def sortedSquares_1(self, nums: list[int]) -> list[int]:
        res = []

        # get the last elem that <= 0 with the binary search
        n = len(nums)
        l, r = 0, n - 1
        while l < r:
            m = (l + r + 1) // 2
            if nums[m] <= 0:
                l = m
            else:
                r = m - 1

        r = l + 1

        for i in range(len(nums)):
            nums[i] = nums[i] ** 2

        # place squares in the right order
        while l > -1 and r < n:
            if nums[l] < nums[r]:
                res.append(nums[l])
                l -= 1
            else:
                res.append(nums[r])
                r += 1

        while l > -1:
            res.append(nums[l])
            l -= 1

        while r < n:
            res.append(nums[r])
            r += 1

        return res

    def sortedSquares_2(self, nums: list[int]) -> list[int]:
        n = len(nums)
        res = [0] * n
        l, r = 0, n - 1
        for i in range(n - 1, -1, -1):
            if abs(nums[l]) > abs(nums[r]):
                res[i] = nums[l] ** 2
                l += 1
            else:
                res[i] = nums[r] ** 2
                r -= 1

        return res
